packages/neon_framework/lib/src/cookie_store/cookie_store.dart

```diff
@@ -0,0 +1,172 @@
+/// > cookie with the earliest last-access date first.
+///
+/// It is recommended to set an upper bound to the time a cookie is stored
+/// as described in [RFC6265 section 7.3](https://httpwg.org/specs/rfc6265.html#rfc.section.7.3):
+/// > Although servers can set the expiration date for cookies to the distant future, most user agents do not actually retain cookies for multiple decades.
+/// > Rather than choosing gratuitously long expiration periods, servers SHOULD promote user privacy by selecting reasonable cookie expiration periods based on the purpose of the cookie.
+/// > For example, a typical session identifier might reasonably be set to expire in two weeks.
+abstract interface class CookieStore {
+  /// Creates a [DefaultCookieStore] instance.
+  factory CookieStore() {
+    return DefaultCookieStore(
+      CookiePersistence(),
+    );
+  }
+
+  /// Save the [cookies] for specified [uri].
+  FutureOr<void> saveFromResponse(Uri uri, List<Cookie> cookies);
+
+  /// Load the cookies for specified [uri].
+  FutureOr<List<Cookie>> loadForRequest(Uri uri);
+
+  /// Ends the current session deleting all session cookies.
+  FutureOr<void> endSession();
+}
+
+/// The default cookie store implementation.
+///
+/// Cookies are parsed into a set of [StorableCookie]s and persisted in the given [persistence].
+final class DefaultCookieStore implements CookieStore {
+  /// Creates a new cookie store backed by the provided [persistence].
+  DefaultCookieStore(this.persistence);
+
+  /// The persistence storing the serialized [StorableCookie].
+  final CookiePersistence persistence;
+
+  @override
+  FutureOr<List<Cookie>> loadForRequest(Uri uri) {
+    return persistence.loadForRequest(uri);
+  }
+
+  @override
+  FutureOr<void> saveFromResponse(Uri uri, List<Cookie> cookies) {
+    final now = tz.TZDateTime.now(tz.UTC);
+    final isHttpRequest = isHttpUri(uri);
+
+    Set<StorableCookie> persistentCookies = {};
+
+    for (final cookie in cookies) {
+      String name = cookie.name;
+      String value = cookie.value;
+      tz.TZDateTime creationTime = now;
+      tz.TZDateTime lastAccessTime = now;
+
+      tz.TZDateTime expiryTime;
+      String domain;
+      String path;
+      bool persistentFlag;
+      bool hostOnlyFlag;
+      bool secureOnlyFlag;
+      bool httpOnlyFlag;
+
+      final maxAge = cookie.maxAge;
+      tz.TZDateTime? expires;
+      if (cookie.expires != null) {
+        expires = tz.TZDateTime.from(cookie.expires!, tz.UTC);
+      }
+
+      if (maxAge != null) {
+        persistentFlag = true;
+        expiryTime = now.add(Duration(seconds: maxAge));
+      } else if (expires != null) {
+        persistentFlag = true;
+        expiryTime = expires;
+      } else {
+        persistentFlag = false;
+        expiryTime = _maxDateTime;
+      }
+
+      // perf: skip already expired cookies.
+      if (!expiryTime.isAfter(now)) {
+        continue;
+      }
+
+      // Skipping public suffix validation
+
```

Comment:
The empty line has to stay.
This is a step reccomended by the RFC but not implemented by us and has nothing to do with the next line.
I'll change the comment to make it more clear.

Reply:
FYI the validation is so we do not accept cookies on the domains that are used publicly by multiple entities.
An example would be the TLD `com` or domains  like `.co.uk`.
We only want to accept cookies from them in case the requestDomain is identical to the public suffix. As otherwise `example.co.uk` could set a cookie on all such co.uk domains that enables cross site tracking without using third party cookies.

Unfortunately we'd need a list of all such suffixes and while the mozilla project maintains one I don't like bloating the package just for this.
I'll add documentation to the class itself mentioning that.

Reply:
Hehe yeah I read some stuff about this problem from the curl developer :D